src/Kestrel.Transport.Sockets/Internal/CoalescingScheduler.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    public class CoalescingScheduler
+    {
+        // Maximum times the work queues swapped and are processed in a single pass
+        private const int _maxLoops = 8;
+
+        private readonly object _workSync = new object();
+        private bool _doingWork;
+        private Queue<Action> _workAdding = new Queue<Action>(1024);
+        private Queue<Action> _workRunning = new Queue<Action>(1024);
+
+        public void Schedule(Action action)
+        {
+            bool scheduleWork;
+
+            lock (_workSync)
+            {
+                _workAdding.Enqueue(action);
+                scheduleWork = !_doingWork;
+                _doingWork = true;
+            }
+
+            if (scheduleWork)
+            {
+                ThreadPool.QueueUserWorkItem(s => ((CoalescingScheduler)s).DoWork(), this);
+            }
+        }
+
+        private void DoWork()
+        {
+            for (var i = 0; i < _maxLoops; i++)
+            {
+                Queue<Action> queue;
+                lock (_workSync)
+                {
+                    queue = _workAdding;
+                    _workAdding = _workRunning;
+                    _workRunning = queue;
+
+                    if (queue.Count == 0)
+                    {
+                        _doingWork = false;
+                        return;
+                    }
+                }
+
+                while (queue.Count != 0)
+                {
+                    queue.Dequeue()();
+                }
+            }
```

Comment:
I'm not clear on why this logic with multiple queues is being used. What are you trying to achieve with it?  Why not something like:
```C#
internal sealed class SingleWorkerScheduler
{
    private static readonly WaitCallback s_worker = s => ((SingleWorkerScheduler)s).ProcessItems();
    private readonly ConcurrentQueue<Action> _actions = new ConcurrentQueue<Action>();
    private bool _hasWorker = false;

    public void Schedule(Action action)
    {
        _actions.Enqueue(action);
        lock (_actions)
        {
            if (!_hasWorker)
            {
                ThreadPool.QueueUserWorkItem(s_worker, this);
                _hasWorker = true;
            }
        }
    }

    private void ProcessItems()
    {
        try
        {
            while (_actions.TryDequeue(out Action item))
            {
                item();
            }
        }
        finally
        {
            lock (_actions)
            {
                if (!_actions.IsEmpty)
                {
                    ThreadPool.QueueUserWorkItem(s_worker, this);
                    _hasWorker = true;
                }
                else
                {
                    _hasWorker = false;
                }
            }
        }
    }
}
```